File: frontend/js/views/perp_funding.js
// Perpetual funding & basis, via /calc/perp-funding.
import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 4 }) + '%');
export async function renderPerpFunding(mount, _s) {
    const tok = currentViewToken(); if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.perpf.h1.title">// PERP FUNDING & BASIS</span></h1>
        <p class="muted small" data-i18n="view.perpf.hint.intro">The cost of carry on a perpetual-futures position and its premium to spot. Funding per interval = notional × funding rate (longs pay shorts when positive); the annualized rate scales by intervals/day over a year; basis = (perp − index) ÷ index. Not financial advice.</p>
        <div class="lpv-split"><div class="chart-panel"><h2 data-i18n="view.perpf.h2.inputs">Inputs</h2>
        <form id="perpf-form" class="inline-form">
            <label><span data-i18n="view.perpf.label.side">Side</span><select name="side"><option value="long" data-i18n="view.perpf.opt.long">Long</option><option value="short" data-i18n="view.perpf.opt.short">Short</option></select></label>
            <label><span data-i18n="view.perpf.label.perp">Perp price ($)</span><input type="number" step="0.01" min="0" name="perp_price_usd" value="30050" required></label>
            <label><span data-i18n="view.perpf.label.index">Index price ($)</span><input type="number" step="0.01" min="0" name="index_price_usd" value="30000" required></label>
            <label><span data-i18n="view.perpf.label.rate">Funding rate/interval (decimal)</span><input type="number" step="0.00001" name="funding_rate" value="0.0001"></label>
            <label><span data-i18n="view.perpf.label.intervals">Intervals/day</span><input type="number" step="1" min="1" name="intervals_per_day" value="3"></label>
            <label><span data-i18n="view.perpf.label.notional">Position notional ($)</span><input type="number" step="100" min="0" name="position_notional_usd" value="100000" required></label>
        </form></div><div id="perpf-result" class="chart-panel lpv-preview"></div></div>`;
    applyUiI18n(mount);
    const form = mount.querySelector('#perpf-form'); const n = (k) => Number(form.querySelector(`[name="${k}"]`).value) || 0;
    const gen = async () => {
        const body = { side: form.querySelector('[name="side"]').value, perp_price_usd: n('perp_price_usd'), index_price_usd: n('index_price_usd'), funding_rate: n('funding_rate'), intervals_per_day: n('intervals_per_day') || 3, position_notional_usd: n('position_notional_usd') };
        try { const d = await api.calcPerpFunding(body); if (!viewIsCurrent(tok)) return; res(mount, d); }
        catch (e) { showToast(e.message || t('view.perpf.toast.error'), { level: 'error' }); }
    };
    const live = debounce(gen, 250); form.addEventListener('input', () => live()); form.addEventListener('submit', (e) => { e.preventDefault(); gen(); }); gen();
}
function res(mount, d) {
    const el = mount.querySelector('#perpf-result');
    const dirKey = d.pays_funding ? 'view.perpf.pays' : 'view.perpf.receives';
    const cls = d.pays_funding ? 'neg' : 'pos';
    el.innerHTML = `<div class="lpv-bar"><div class="cards">
        <div class="card ${cls}"><div class="label" data-i18n="view.perpf.card.dir">Funding</div><div class="value" data-i18n="${dirKey}">${d.pays_funding ? 'Pays' : 'Receives'}</div></div>
        <div class="card"><div class="label" data-i18n="view.perpf.card.interval">Per interval</div><div class="value">${money(d.funding_per_interval_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.perpf.card.day">Per day</div><div class="value">${money(d.funding_per_day_usd)}</div></div>
        <div class="card"><div class="label" data-i18n="view.perpf.card.ann">Annualized</div><div class="value">${pct(d.annualized_funding_pct)}</div></div>
        <div class="card"><div class="label" data-i18n="view.perpf.card.basis">Basis</div><div class="value">${pct(d.basis_pct)}</div></div>
    </div></div>`;
    applyUiI18n(el);
}

File: crates/traderview-core/src/perp_funding.rs
//! Perpetual funding & basis — the cost of carry on a crypto perpetual-futures
//! position and its premium to spot. The funding payment per interval is
//! `notional × funding_rate` (longs pay shorts when the rate is positive); the
//! annualized funding rate scales it by the intervals per day over a year; and
//! the basis is the perp's premium over the index, `(perp − index) ÷ index`.
//! Reports who pays and the position's daily and annualized funding cost. Pure
//! compute. Not financial advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct PerpFundingInput {
    pub perp_price_usd: f64,
    pub index_price_usd: f64,
    /// Funding rate per interval, decimal (e.g. 0.0001 = 0.01%).
    pub funding_rate: f64,
    /// Funding intervals per day (e.g. 3 for 8-hour funding).
    #[serde(default = "default_intervals")]
    pub intervals_per_day: f64,
    /// Position notional in USD.
    pub position_notional_usd: f64,
    /// "long" or "short".
    pub side: String,
}

fn default_intervals() -> f64 {
    3.0
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct PerpFundingReport {
    /// (perp − index) ÷ index, percent.
    pub basis_pct: f64,
    /// Funding paid (+) or received (−) per interval for this position's side.
    pub funding_per_interval_usd: f64,
    pub funding_per_day_usd: f64,
    /// funding_rate × intervals_per_day × 365, percent.
    pub annualized_funding_pct: f64,
    /// True when this side pays funding (long & rate>0, or short & rate<0).
    pub pays_funding: bool,
}

fn round2(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

pub fn generate(i: &PerpFundingInput) -> PerpFundingReport {
    let basis = if i.index_price_usd > 0.0 {
        (i.perp_price_usd - i.index_price_usd) / i.index_price_usd * 100.0
    } else {
        0.0
    };
    let is_long = !i.side.trim().eq_ignore_ascii_case("short");
    // Longs pay when the rate is positive; shorts pay when negative.
    let pays = (is_long && i.funding_rate > 0.0) || (!is_long && i.funding_rate < 0.0);
    // Signed magnitude: positive = this side pays out.
    let direction = if is_long { 1.0 } else { -1.0 };
    let per_interval = i.position_notional_usd * i.funding_rate * direction;
    let per_day = per_interval * i.intervals_per_day;
    let annualized = i.funding_rate * i.intervals_per_day * 365.0 * 100.0;

    PerpFundingReport {
        basis_pct: round4(basis),
        funding_per_interval_usd: round2(per_interval),
        funding_per_day_usd: round2(per_day),
        annualized_funding_pct: round2(annualized),
        pays_funding: pays,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> PerpFundingInput {
        PerpFundingInput {
            perp_price_usd: 30_050.0,
            index_price_usd: 30_000.0,
            funding_rate: 0.0001,
            intervals_per_day: 3.0,
            position_notional_usd: 100_000.0,
            side: "long".into(),
        }
    }

    #[test]
    fn long_pays_positive_funding() {
        let d = generate(&base());
        assert!(close(d.basis_pct, 0.1667));
        assert!(close(d.funding_per_interval_usd, 10.0));
        assert!(close(d.funding_per_day_usd, 30.0));
        assert!(close(d.annualized_funding_pct, 10.95));
        assert!(d.pays_funding);
    }

    #[test]
    fn short_receives_positive_funding() {
        let d = generate(&PerpFundingInput { side: "short".into(), ..base() });
        // Short receives → negative payment.
        assert!(close(d.funding_per_interval_usd, -10.0));
        assert!(!d.pays_funding);
    }

    #[test]
    fn negative_funding_long_receives() {
        let d = generate(&PerpFundingInput { funding_rate: -0.0001, ..base() });
        assert!(close(d.funding_per_interval_usd, -10.0));
        assert!(!d.pays_funding);
    }

    #[test]
    fn discount_basis_negative() {
        let d = generate(&PerpFundingInput { perp_price_usd: 29_900.0, ..base() });
        assert!(d.basis_pct < 0.0);
    }
}
